cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/ConsumerWorkingSet.java

```diff
@@ -0,0 +1,171 @@
+        return partitions.remove(i);
+      }
+    }
+    throw new IllegalArgumentException("PartitionKey not found: " + partitionKey);
+  }
+
+  public ConsumablePartition lookup(PartitionKey partitionKey) {
+    for (int i = 0; i < partitions.size(); i++) {
+      if (partitionKey.equals(partitions.get(i).getPartitionKey())) {
+        return partitions.get(i);
+      }
+    }
+    throw new IllegalArgumentException("PartitionKey not found: " + partitionKey);
+  }
+
+  public void setPartitionConsumerState(PartitionConsumerState partitionConsumerState) {
+    this.partitionConsumerState = partitionConsumerState;
+  }
+
+  public PartitionConsumerState getPartitionConsumerState() {
+    return partitionConsumerState;
+  }
+
+  public static ConsumerWorkingSet fromBytes(byte[] bytes) {
+    ByteBuffer bb = ByteBuffer.wrap(bytes);
+    byte serializationFormatVersion = bb.get();
+    if (serializationFormatVersion != 0) {
+      throw new IllegalArgumentException("Unsupported serialization format: " + serializationFormatVersion);
+    }
+
+    int numPartitions = bb.getInt();
+    List<ConsumablePartition> partitions = new ArrayList<>(numPartitions);
+    for (int i = 0; i < numPartitions; i++) {
+      ProcessState processState = ProcessState.fromByte(bb.get());
+      int keyLength = bb.getInt();
+      byte[] stringBytes = new byte[keyLength];
+      bb.get(stringBytes, 0, keyLength);
+      long timestamp = bb.getLong();
+      partitions.add(new ConsumablePartition(GSON.fromJson(Bytes.toString(stringBytes), PartitionKey.class),
+                                             processState, timestamp));
+    }
+
+    int sizeOfMarker = bb.getInt();
+    byte[] markerBytes = new byte[sizeOfMarker];
+    bb.get(markerBytes);
+    return new ConsumerWorkingSet(PartitionConsumerState.fromBytes(markerBytes), partitions);
+  }
+
+  public byte[] toBytes() {
+    // first byte for serialization format version
+    int numBytes = 1;
+    numBytes += Ints.BYTES;
+    for (ConsumablePartition partition : partitions) {
+      // 1 byte for the ProcessState
+      numBytes += 1;
+      numBytes += Ints.BYTES;
+      numBytes += partition.getPartitionKeyBytes().length;
+      numBytes += Longs.BYTES;
+    }
+    byte[] markerBytes = partitionConsumerState.toBytes();
+    numBytes += Ints.BYTES;
+    numBytes += markerBytes.length;
+
+    ByteBuffer bb = ByteBuffer.allocate(numBytes);
+    // currently, serialization format is 0
+    bb.put((byte) 0);
+    bb.putInt(partitions.size());
+    for (ConsumablePartition partition : partitions) {
+      byte[] partitionKeyBytes = partition.getPartitionKeyBytes();
+      bb.put(partition.getProcessState().toByte());
+      bb.putInt(partitionKeyBytes.length);
+      bb.put(partitionKeyBytes);
+      bb.putLong(partition.getTimestamp());
+    }
+    bb.putInt(markerBytes.length);
+    bb.put(markerBytes);
+    return bb.array();
+  }
+
+  @Override
```

Comment:
is equals and hashcode used for this? If not, better not implement them. they easily get out of date when a field is added.